### PR TITLE
Add support for SmartAsserts.jl

### PR DESCRIPTION
**Pull request by Zentrik**
It should be easy to add any other similar package such as ArgChecks.jl

### CODE REVIEW

**Comment by mauro3** [comment]
Thanks!

Comments:
- adding this feature would be good
- shouldn't ArgCheck.jl be implemented as on https://github.com/MrVPlusOne/SmartAsserts.jl it states that that package is preferred?
- The tests don't actually test the functionality or am I missing something?

x-ref: #120

**Comment by Zentrik** [comment]
I decided not to use ArgCheck as it's ability to compile out asserts is not supported on new versions of julia, [OptionalArgChecks.jl](https://github.com/simeonschaub/OptionalArgChecks.jl) is broken on >= 1.5.
I would have added ArgCheck as well in this pr but I couldn't figure out a way to add the test without just copy and pasting them, which I didn't want to do as otherwise you would have 3 copies of near identical tests. If you don't care about this I can add that ArgCheck as well.

I've updated the tests, thanks for spotting that.

The main problem now is I didn't notice SmartAssert has set its compat to julia 1.5, I'll see if I can get that changed.

**Comment by Zentrik** [comment]
This should be ready to be merged in now.

File: Manifest.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (mauro3) [inline Manifest.toml:1]: This file should not be included in the PR

File: Project.toml [hunk 2]
```diff
@@ -14,8 +14,9 @@ julia = "1"
```
Comment (mauro3) [inline Project.toml:14]: Let's just move compat to Julia 1.5

File: Project.toml [hunk 3]
```diff
@@ -14,8 +14,9 @@ julia = "1"
 
 [extras]
 Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
```
Comment (mauro3) [inline Project.toml:17]: Shouldn't you just add `SmartAsserts` here and drop `Pkg`?  Then you can just use it in the tests.

File: test/runtests.jl [hunk 4]
```diff
@@ -378,6 +384,36 @@ end
 @test_throws AssertionError MT12a([1,2])
 @test MT12a([1]).a==MT12a(a=[1]).a
 
+### Smart Assertions
+
+@static if VERSION >= v"1.5" # Need static otherwise it tries to evaluate code below which fails
```
Comment (mauro3) [inline test/runtests.jl:389]: Dropping Julia 1.0 support will remove this.

**Review by mauro3 — CHANGES_REQUESTED** [review]
Thanks for adding the tests!  There are still a few bits to do.  Also, let's move the compat for Julia to 1.5.

**Review by mauro3 — APPROVED** [review]

**Comment by mauro3** [comment]
Sorry, one more thing: now we just need some docs!

**Comment by mauro3** [comment]
Thanks so much for your help!

**Pull request merged**